Self-service site restore: update UI

mpkelly opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/6287 
Related to D143388-code
Related to D143193-code

<img width="216" alt="Screenshot 2567-03-28 at 12 44 08" src="https://github.com/Automattic/wp-calypso/assets/6851384/4cd734fe-e919-4115-b1c0-3e025be2887c">


<img width="1116" alt="Screenshot 2567-04-01 at 17 41 52" src="https://github.com/Automattic/wp-calypso/assets/6851384/8253565a-8f3c-4cdd-a451-d7a9de9e6707">

<img width="388" alt="Screenshot 2567-04-01 at 17 41 39" src="https://github.com/Automattic/wp-calypso/assets/6851384/30f2400b-8314-4c95-b31a-49566365e78d">


## Proposed Changes

* Allow accessing and restoring deleted sites at `/sites` when the status **Deleted** is selected. Deleted sites are not returned for any other status. 

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Delete a site
* Go to /sites
* Filter to "Deleted"
* Restore the site
* Filter to "All"
* See the site is back and that you're admin 
* Try it with multiple users. They all should be added back. 

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

mpkelly committed:
GitHub deployments: remove client-side templates and use template name instead

mpkelly committed:
Self-service site restore: update UI (WIP)

mpkelly committed:
Fetch deleted sites and support filtering

mpkelly committed:
Support restoring a site via new endpoint.

mpkelly commented:
@javierarce, thanks. That makes total sense and is easier to code and cleaner than what I did. I will update it today.

zaguiini commented on client/sites-dashboard/components/sites-dashboard.tsx:
```diff
@@ -176,10 +176,21 @@ export function SitesDashboard( {
+		'deleted'
+	);
+
+	deletedSites.forEach( ( site ) => {
+		site.is_deleted = true;
+	} );
```
We shouldn't make this change during render. [We could probably use the select property from react-query](https://tkdodo.eu/blog/react-query-data-transformations).

zaguiini commented on client/sites-dashboard/components/sites-table-row.tsx:
```diff
@@ -167,19 +178,57 @@ export default memo( function SitesTableRow( { site }: SiteTableRowProps ) {
 		siteUrl = site.options?.unmapped_url;
 	}
 
+	const isDeleted = site.is_deleted;
```
Do we need this extra variable?

zaguiini commented on client/sites-dashboard/components/sites-table-row.tsx:
```diff
@@ -205,53 +258,77 @@ export default memo( function SitesTableRow( { site }: SiteTableRowProps ) {
+			{ isDeleted ? (
+				<Column style={ { textAlign: 'right' } }>
+					<Button scary busy={ isPending } disabled={ isPending } onClick={ handleRestoreSite }>
+						{ __( 'Restore site ' ) }
```
Would be good if we could trim the translatable strings:

```suggestion
						{ __( 'Restore site' ) }
```

zaguiini commented on client/sites-dashboard/components/sites-table.tsx:
```diff
@@ -137,6 +137,8 @@ export function SitesTable( { className, sites, isLoading = false }: SitesTableP
 		};
 	}, [ masterbarHeight ] );
 
+	const isStatusDeleted = sites.some( ( site ) => ! site.is_deleted );
```
This probably deserves it `useMemo` wrapping so that we can avoid calculating this on every render. It's ok for smaller samples but once there are a lot of sites it becomes a bottleneck.

mpkelly replied:
It's been removed after Javier's feedback. We will keep the table columns the same, so we don't need this.

zaguiini commented on client/sites-dashboard/components/sites-table.tsx:
```diff
@@ -146,15 +148,17 @@ export function SitesTable( { className, sites, isLoading = false }: SitesTableP
+							</StatsThInner>
+						</StatsTh>
+					) }
+					<th style={ isStatusDeleted ? {} : { width: '24px' } }></th>
```
```suggestion
					<th style={ isStatusDeleted ? undefined : { width: '24px' } }></th>
```

zaguiini reviewed (commented):
Left some code review notes

mpkelly committed:
Restore old site table

mpkelly committed:
Rework table after feedback

mpkelly committed:
Unrelated to PR: fix wonky "Stats off" label

mpkelly committed:
Fix tests

mpkelly commented:
Thanks for the feedback, @javierarce and @zaguiini. 

I've updated the UI so it looks more like a regular row. Some notes:

- There is no plan for a deleted site, so is empty. 
- The padding around the Restore button text is a bit tight (see below). I can adjust it but it will throw the column alignment off. 

<img width="177" alt="Screenshot 2567-03-28 at 12 39 59" src="https://github.com/Automattic/wp-calypso/assets/6851384/d4532881-3b85-4072-a066-5a97912edf71">

- I would like some feedback about "All". Deleted sites are only shown when you choose "Deleted". Do you agree with this?
- A deleted status option appears on the migration flow, but it will always be empty because we are not fetching deleted sites on this flow. It seems harmless but I can remove this option on this flow is you want. 

<img width="1051" alt="Screenshot 2567-03-28 at 12 35 38" src="https://github.com/Automattic/wp-calypso/assets/6851384/aff9c34a-aba6-49c6-8756-d11b7cc2d7cb">

mpkelly committed:
Fix types

vykes-mac commented:
> I would like some feedback about "All". Deleted sites are only shown when you choose "Deleted". Do you agree with this, or should they be included with "All"?

I personally don't think they should appear under `all`, deleted sites to me is like trash can or archived, It's there but doesn't appear under the normal folders. Most systems work this way so this would be something users are familiar with.

> A deleted status option appears on the migration flow, but it will always be empty because we are not fetching deleted sites on this flow. It seems harmless but I can remove this option on this flow is you want. 

I agree this seems harmless to me

vykes-mac commented:
I think we need to add a note to the deleted sites page letting users know the sites will be permanently removed after 30 days

vykes-mac committed:
add a notice to let users know deleted sites will be permanently removed

vykes-mac reviewed (approved):
I'm good with it!

javierarce commented:
I found a small issue. If the user has the "Classic Dark" theme activated, it's very difficult to distinguish between the two colors.

<img width="1438" alt="Screenshot 2024-03-29 at 18 23 26" src="https://github.com/Automattic/wp-calypso/assets/4933/213f6cdc-908f-4ad7-9948-91d1ffed461a">

Instead of using `var(--color-accent)`, let's follow the style of the "Upgrade" link and use `var(--color-link)`. And let's also set it in 12px with an underline. 

So we would essentially use the same style as the "Upgrade" link and be consistent with that design instead of the one from the Launch checklist. Sorry, I only realize now that this change is better than the one I proposed originally.

<img width="1127" alt="Screenshot 2024-03-29 at 18 27 45" src="https://github.com/Automattic/wp-calypso/assets/4933/c9a190b4-5959-4495-8034-d70398fec118">

javierarce commented:
> I would like some feedback about "All". Deleted sites are only shown when you choose "Deleted". Do you agree with this, or should they be included with "All"?

I guess we could agree that a deleted site isn't technically a site, so in that case, it would be correct to keep them out of the list. My only concern is that by not displaying them, we're going to make the restore option difficult to find because the default state of the dropdown gives the impression that it's displaying everything. If we are okay with that, I'd say not to include them… otherwise, we'll probably need to find an alternative.

> A deleted status option appears on the migration flow, but it will always be empty because we are not fetching deleted sites on this flow. It seems harmless but I can remove this option on this flow is you want.

If we can remove it from there, let's do it, because even if people won't select that, it doesn't make sense to display that option in that context.

mpkelly committed:
Fix mobile layout

mpkelly committed:
Merge branch 'trunk' into github-deployments/allow-restoring-a-deleted-site

mpkelly committed:
Fix build

mpkelly committed:
Update link colour as per design feedback

mpkelly committed:
Only show deleted status at /sites, other locations

mpkelly commented:
Thanks, @javierarce. I changed the link styling and removed the "Deleted" status from the import screen.

mpkelly committed:
Fix type error

javierarce commented:
> I guess we could agree that a deleted site isn't technically a site, so in that case, it would be correct to keep them out of the list. My only concern is that by not displaying them, we're going to make the restore option difficult to find because the default state of the dropdown gives the impression that it's displaying everything. If we are okay with that, I'd say not to include them… otherwise, we'll probably need to find an alternative.

I think I found a nice workaround to help with the discoverability of the deleted option in the dropdown: we can point users to the `/sites?status=deleted` in the https://github.com/Automattic/dotcom-forge/issues/5103. It's not perfect, but at least we can inform users of the existence of that option when it's more relevant to them.

mpkelly committed:
Fix type error

mpkelly committed:
Don't show link if deleted

mpkelly committed:
Show deleted with "All"

mpkelly committed:
Seem to have picked this up with a rebase but it shouldn't be in this PR. Reverting.
